README.md
# Circonus API Client Library

The `circonusapi` python module contains three classes:

- CirconusAPI

- CirconusData

- CirconusSubmit

The CirconusAPI class contains methods to manage the Circonus Account (e.g. create checks, graphs, etc).
The CirconusData class contains higher-level methods for fetching data. 
In particular it comes with a method that returns CAQL results as Pandas DataFrames.
The CirconusSubmit class contains methods for submitting data to Circonus via a JSON HTTPTrap.

## Installation

via pip

    pip install circonusapi
   
Manual Install

    python setup.py install

## Requirements and Dependencies

- CirconusAPI supports Python 2.6, 2.7 and 3.x

- CirconusSubmit and CirconusData are tested against python 3.x

**Optional Dependencies:**

* Histogram functionality depends on [libcircllhist](github.com/circonus-labs/libcircllhist) being installed.

* The method `CirconusData.caqldf()` depends on [pandas](https://pandas.pydata.org/) being installed.


docs/changelog.rst
Changelog
=========

v0.6.0
  - Added experimental ./bin/caql cli tool
  - Support http/https connections to API endpoints
  - Add Sphinx based documentation at gh-pages
  - Support test-docker target, that runs tests with all supported python versions

v0.5.0
  * Date: 2020-02-07
  * Changes:

    - Add CirconusSubmit class

v0.4.0
  * Date: 2020-01-17
  * Changes:

    - Add metadata and DF4 support to CirconusData.caql()
    - Add CirconusData.caqldf() method
    - Remove CirconusData.search() method. Search functionality is provided via CirconusData.caql().
    - Deprecate config.py module


test/run.sh
set -e

export CIRCONUS_CONFIG=./test_config.ini

python test_circonusapi.py

if python -c 'import sys; sys.exit(not sys.version_info.major == 3 )'
then
  # python3 only tests
  python test_circonusdata.py
fi


test_circonusapi.py
'''
To test actual endpoints create a valid config file and point to it.

Example:

CIRCONUS_CONFIG=/tmp/test_circonus_config.ini tox

The config file must contain a valid token for the Circonus demo account.
'''
import os

from tempfile import NamedTemporaryFile
import unittest
from unittest import TestCase

from circonusapi import circonusapi, config

class ConfigTestCase(TestCase):

    def test_no_config_file(self):
        cfg = config.load_config(configfile='nope', nocache=True)
        self.assertEqual(cfg.sections(), [])

    def test_single_config_file(self):
        with NamedTemporaryFile(mode='w+') as tmp_file:
            tmp_file.file.write(
                '''
[general]
default_account=sampleaccount
appname=sample appname

[tokens]
sampleaccount=AAAAAAAA-1234-5678-9999-BBBBBBBB
'''
            )
            tmp_file.file.flush()
            tmp_file.file.seek(0)
            cfg = config.load_config(
                configfile=tmp_file.name,
                nocache=True
            )
            self.assertEqual(cfg.get('general', 'appname'), 'sample appname')
            cfg_cached = config.load_config()
            self.assertEqual(cfg_cached.get('general', 'appname'), 'sample appname')


    def test_two_merged_config_files(self):
        config_txt_1 = '''
[general]
default_account=sampleaccount
appname=sample appname
'''
        config_txt_2 = '''
[tokens]
sampleaccount=AAAAAAAA-1234-5678-9999-BBBBBBBB
'''
        with NamedTemporaryFile(mode='w+') as tmp_file_1:
            with NamedTemporaryFile(mode='w+') as tmp_file_2:
                tmp_file_1.write(config_txt_1)
                tmp_file_2.write(config_txt_2)
                for cfl in [tmp_file_1, tmp_file_2]:
                    cfl.file.flush()
                    cfl.file.seek(0)
                cfg = config.load_config(
                    configfile=[
                        tmp_file_1.name,
                        tmp_file_2.name
                    ],
                    nocache=True
                )
            self.assertEqual(cfg.get('general', 'appname'), 'sample appname')
            self.assertEqual(cfg.get('tokens', 'sampleaccount'), 'AAAAAAAA-1234-5678-9999-BBBBBBBB')


class CirconusAPITestCase(TestCase):

    def setUp(self):
        config_file = os.environ.get('CIRCONUS_CONFIG')
        cfg = config.load_config(configfile=config_file, nocache=True)
        account = cfg.get('general', 'default_account')
        appname = cfg.get('general', 'appname')
        token = cfg.get('tokens', account)
        api = circonusapi.CirconusAPI(token)
        api.appname = appname
        self.config = cfg
        self.api = api

    def test_brokers(self):
        brokers = self.api.list_broker()
        self.assertTrue(len(brokers) > 10)

    def test_wrong_method(self):
        try:
            result = self.api.list_foobar()
        except AttributeError:
            pass

    def test_data_endpoint(self):
        result = self.api.get_data(
            '160764_services',
            params=dict(
                type='numeric',
                start=1514764800,
                end=1514767800,
                period=300
            )
        )
        self.assertTrue(len(result.get('data')) >= 10)

    def test_data_wrong_format(self):
        '''
        Test 'data' endpoint and whether errors properly deserialized.
        '''
        try:
            result = self.api.get_data(123)
        except circonusapi.CirconusAPIError as e:
            self.assertEqual(e.code, 400)

    def test_contact_group(self):
        contacts = self.api.list_contact_group()
        self.assertTrue(type(contacts) is list)

    def test_search_dns_bundles(self):
        dns_bundles = self.api.list_check_bundle(
            params=dict(f_type='dns')
        )
        for bundle in dns_bundles:
            self.assertEqual(bundle.get('type'), 'dns')

if __name__ == '__main__':
    unittest.main()


test_circonusdata.py
"""
Test for the circonusdata module
"""
#
# circonusdata.py depends on circllhist being available. As this dependency is currently not
# installable via pip we keep these tests in a separate file.
#
import os
from datetime import datetime

import unittest
from unittest import TestCase

from circonusapi import circonusdata, config

class CirconusAPITestCase(TestCase):

    def setUp(self):
        config_file = os.environ.get('CIRCONUS_CONFIG')
        cfg = config.load_config(configfile=config_file, nocache=True)
        account = cfg.get('general', 'default_account')
        token = cfg.get('tokens', account)
        api = circonusdata.CirconusData(token)
        self.api = api

    def test_search(self):
        m = self.api.search("(check_id:160764) (metric:services)")
        self.assertEqual(len(m), 1)
        r = m.fetch(datetime(2018,1,1),60,10)
        self.assertEqual(len(r['time']), 10)

    def test_caql(self):
        c = self.api.caql("search:metric('(check_id:160764) (metric:services)')", datetime(2018,1,1), 60, 10)
        self.assertEqual(len(c['time']), 10)

if __name__ == '__main__':
    unittest.main()


searchindex.js
Search.setIndex({docnames:["api","changelog","data","index","submit"],envversion:{"sphinx.domains.c":1,"sphinx.domains.changeset":1,"sphinx.domains.citation":1,"sphinx.domains.cpp":1,"sphinx.domains.index":1,"sphinx.domains.javascript":1,"sphinx.domains.math":2,"sphinx.domains.python":1,"sphinx.domains.rst":1,"sphinx.domains.std":1,sphinx:56},filenames:["api.rst","changelog.rst","data.rst","index.rst","submit.rst"],objects:{"circonusapi.circonusapi":{CirconusAPI:[0,1,1,""]},"circonusapi.circonusapi.CirconusAPI":{api_call:[0,2,1,""]},"circonusapi.circonusdata":{CirconusData:[2,1,1,""]},"circonusapi.circonusdata.CirconusData":{caql:[2,2,1,""],caqldf:[2,2,1,""],from_api:[2,2,1,""],from_irondb:[2,2,1,""]},"circonusapi.circonussubmit":{CirconusSubmit:[4,1,1,""]},"circonusapi.circonussubmit.CirconusSubmit":{add_histogram:[4,2,1,""],add_number:[4,2,1,""],auth:[4,2,1,""],check_create:[4,2,1,""],submit:[4,2,1,""]},circonusapi:{circonusapi:[0,0,0,"-"],circonusdata:[2,0,0,"-"],circonussubmit:[4,0,0,"-"]}},objnames:{"0":["py","module","Python module"],"1":["py","class","Python class"],"2":["py","method","Python method"]},objtypes:{"0":"py:module","1":"py:class","2":"py:method"},terms:{"4ede":4,"65669d6b":4,"7f3cae6419cf":4,"boolean":2,"class":[1,3],"function":[1,2],"import":[0,2,4],"int":2,"new":4,"null":0,"return":[2,3],"true":2,Added:1,And:0,For:0,The:[0,2,3],Use:[2,4],__check_target:2,_last_modified_bi:0,absenc:0,access:2,accordingli:4,account:[0,2,3],action:0,add:[0,1,4],add_histogram:4,add_numb:4,add_rule_set:0,add_x:0,against:2,all:[0,1,2],also:0,altern:0,api:[0,1,2,4],api_cal:0,api_token:[0,2],appnam:0,appropri:0,arg:[0,2,4],associ:0,auth:4,authent:4,avail:2,avoid:2,back:0,base:1,baseurl:0,batch:4,bc51:4,beforehand:4,begin:0,bin:1,broker:0,bundl:0,call:[2,4],can:[0,2],caql:[1,2,3],caqldf:[1,2],chang:1,check:[0,3,4],check_bundl:0,check_creat:4,circ:[2,4],circllhist:[2,4],circonu:[0,2,3,4],circonusapi:[2,4],circonusdata:[1,3],circonussubmit:[1,3],classmethod:2,cli:1,column:2,com:[0,2],come:3,config:1,connect:[1,2],constructor:2,contact_group:0,contain:[0,3],convert:2,convert_hist:2,count:2,creat:[2,3,4],criteria:0,current:[0,2],data:[0,2,3,4],datafram:[2,3],datapoint:2,date:1,datetim:[2,4],debug:0,delet:0,deprec:1,deriv:0,dev:2,df4:[1,2],dict:[0,2],direct:2,directli:0,docker:1,document:1,durat:2,edf:4,edit:0,either:[0,2],endpoint:[0,1,2],epoch:4,especi:0,etc:3,exist:4,experiment:1,explain:2,factori:2,fals:[0,2],fetch:[0,2,3],find:[0,2],folder:2,foo:0,form:[0,2],format:2,found:2,from:[0,2,4],from_api:2,from_irondb:2,full:2,gain:2,get:0,get_check_bundl:0,get_us:0,given:4,graph:[0,3],have:0,head:2,higher:3,hist:4,histogram:[2,4],hostnam:2,http:[0,1,2],httptrap:[3,4],implement:2,includ:4,index:2,initi:0,instanc:2,instead:2,irondb1:2,irondb:2,isn:0,issu:2,json:[0,3],k8sdemo2:2,kind:2,kwarg:2,label:2,level:[0,3],limit:2,link:0,list:0,list_check_bundl:0,list_rule_set:0,list_x:0,localhost:2,lower:0,make:0,manag:3,meta:2,metadata:[1,2],method:[1,2,3,4],metric:[2,4],metric_nam:0,metric_typ:0,modifi:0,modul:[1,2,3],more:2,my_result:0,name:[2,4],net:2,next:4,node:2,none:[0,2,4],note:[0,2],now:4,number:[0,2,4],numer:[0,2,4],object:[2,4],oper:[0,2],option:[2,4],out:0,output:2,page:1,panda:[2,3],param:0,parent:0,pars:0,particular:3,payload:0,per:2,perform:0,period:2,port:2,post:0,protocol:2,provid:[1,2,4],put:0,python:[0,1,3],queri:[0,2],referenc:0,regular:0,remov:1,repositori:2,request:[0,2],requir:[0,2,4],res:2,resource_id:0,result:[2,3],row:2,rule:0,rule_set:0,run:[0,1,2],search:1,second:[2,4],send:0,set:[0,4],sever:0,should:2,simplifi:2,sinc:4,singl:2,some_metr:2,specif:0,specifi:0,sphinx:1,start:2,starttim:2,str:[0,2,4],stream:4,string:[0,2,4],sub:4,submiss:4,submit:[3,4],support:1,tag:4,take:0,target:1,templat:0,test:[1,4],thi:[0,2,4],three:3,timestamp:[2,4],token:[0,2,4],tool:1,two:0,underscor:0,unix:2,url:[2,4],use:[0,2],useful:0,user:0,using:[0,2],valid:0,valu:[0,4],verb:0,version:[1,2],via:[1,3,4],view:0,wait:0,whatev:0,when:0,where:0,who:0,wish:0,without:0,would:0,xkcd:2,xyz:0,you:0},titles:["Class CirconusAPI","Changelog","Class CirconusData","Welcome to Circonusapi","Class CirconusSubmit"],titleterms:{"class":[0,2,4],call:0,changelog:1,circonusapi:[0,3],circonusdata:2,circonussubmit:4,direct:0,exampl:[0,2,4],method:0,paramet:0,welcom:3}})

circonusapi/util.py
"""Various helper functions to make using the API a little easier

Many functions come in a regular and 'pretty' version. The pretty version is
designed to be used in an interactive setting, handling errors, printing
messages and asking for input from the command line. The regular version
doesn't deal with errors, and avoids printing messages where possible.
"""
import log
import sys
import os
import re
import json
import config

def confirm(text="OK to continue?"):
    response = None
    while response not in ['Y', 'y', 'N', 'n']:
        response = raw_input("%s (y/n) " % text)
    if response in ['Y', 'y']:
        return True
    return False

def get_broker(api, broker_name):
    """Find a broker endpoint given its name"""
    rv = api.list_broker()
    brokers = dict([(i['name'], i['_cid']) for i in rv])
    return brokers[broker_name]

def get_broker_pretty(api, broker_name):
    """Find a broker endpoint given its name and exit with an error if it's
    not found.
    """
    try:
        return get_broker(api, broker_name)
    except KeyError:
        log.error("Invalid/Unknown Broker: %s" % broker_name)
        sys.exit(1)

def find_check_bundle(api, pattern):
    """Searches for check_bundles via regular expression on the check name

    Returns a list of checks, and also the values of any matching groups
    specified in the regex.
    """
    all_bundles = api.list_check_bundle()
    filtered_bundles = []
    groups = {}
    for b in sorted(all_bundles):
        m = re.search(pattern, b['display_name'])
        if m:
            filtered_bundles.append(b)
            # Store numbered groups
            matchgroups = m.groups()
            groups[b['_cid']] = {}
            for i in range(0, len(matchgroups)):
                groups[b['_cid']]["group%s" % (i + 1)] = matchgroups[i]
            # Store named groups - (?P<name>...)
            groups[b['_cid']].update(m.groupdict())
    return {
        'bundles': filtered_bundles,
        'groups': groups
    }

def find_check_bundle_pretty(api, pattern):
    log.msg("Retrieving matching checks")
    return find_check_bundle(api, pattern)

def find_metrics(check_bundle, pattern):
    """Retreives metrics for a check bundle by regex

    Also returns a list of metrics that didn't match. Useful when disabling
    metrics - where you need to include all metrics except those that matched.

    Parameters:

        check_bundle - a check bundle returned from the API
        pattern - the regex to match metrics on
    """
    all_metrics = check_bundle['metrics']
    matching_metrics = []
    non_matching_metrics = []
    for i in sorted(all_metrics):
        m = re.search(pattern, i['name'])
        if m:
            matching_metrics.append(i)
        else:
            non_matching_metrics.append(i)
    return {
        'matching': matching_metrics,
        'non_matching': non_matching_metrics
    }

def verify_metrics_pretty(template, check_bundles):
    log.msg("Verifying that bundles have the correct metrics")
    template_metrics = template.get_metrics()
    bundles_with_correct_metrics = []
    bundles_with_wrong_metrics = []
    count = 0
    for b in check_bundles:
        count += 1
        print "\r%s/%s" % (count, len(check_bundles)),
        sys.stdout.flush()
        metrics = b['metrics']
        metric_names = [m['name'] for m in metrics]
        for m in template_metrics:
            if m not in metric_names:
                bundles_with_wrong_metrics.append({
                    'name': b['display_name'],
                    'metric': m})
            else:
                bundles_with_correct_metrics.append(b)
    if bundles_with_wrong_metrics:
        log.msg("The following check bundles do not have metrics specified in"
                " the template:")
        for c in bundles_with_wrong_metrics:
            log.msg("%(name)s - %(metric)s" % c)
        if confirm("Do you want to continue with just the check bundles that"
                " match the template?"):
            log.msg("Continuing with only matching check bundles")
        else:
            log.error("Not continuing. The template does not match the"
                    " bundles")
            sys.exit(1)
        return bundles_with_correct_metrics



circonusapi/circonusdata.py
"""
==================
Class CirconusData
==================

This module provides simplified methods for data fetching operations.
To gain access to the full functionality use the circonusapi module.

Example
=======

::

    from circonusapi import circonusdata

    # Option A: Connect to the Circonus API using a token
    circ = circonusdata.from_api(api_token)

    # Option B: Connect to IRONdb instance
    # circ = circonusdata.from_irondb("http://irondb1.dev.net:8112", account=27)

    # Run a CAQL query
    from datetime import datetime
    circ.caql('''

      1 + 2 | label("A")

    ''', datetime(2020,1,1), 60, 10)

    # Result
    #
    # {
    #   'version': 'DF4',
    #    'head': {'count': 10, 'start': 1577836800, 'period': 60},
    #    'meta': [{'kind': 'numeric', 'label': 'A'}],
    #    'data': [[3, 3, 3, 3, 3, 3, 3, 3, 3, 3]]
    # }

    # Fetch CAQL as DataFrame
    circ.caqldf('''

        find("duration", limit=10) | label("%tv{__check_target}")

    ''', datetime(2020, 1, 1), 60, 60 * 4)

    # Result
    #
    #                      xkcd.com  xkcd.com  151.101.64.67  k8sdemo2
    # 2020-01-01 00:00:00         3         3              3        49
    # 2020-01-01 00:01:00         3         3              3        55
    # 2020-01-01 00:02:00         3         3              3        61
    # 2020-01-01 00:03:00         3         3              3        48


More examples can be found in the ./examples folder in this repository.

"""

import math
from datetime import datetime
import warnings
import requests

from . import circonusapi

#
# Optional Imports
#

try:
    from circllhist import Circllhist
except ImportError:
    Circllhist = None

try:
    import pandas as pd
except ImportError:
    pd = None


class CirconusData(object):
    """Circonus data fetching class.

    Direct constructor calls should be avoided.
    Use the provided factory methods .from_api() / .from_irondb() to create instances of this class.
    """


    def __init__(self, token=None, endpoint=None, account=1):
        if token:
            self._mode = "API"
            self._api = circonusapi.CirconusAPI(token)
        elif endpoint:
            self._mode = "IRONdb"
            self._endpoint = endpoint
            self._account = account
        else:
            raise Exception("No token/endpoint given")

    @classmethod
    def from_api(cls, token):
        """
        Connect to the Circonus API with a token

        Args:
           token (str): Circonus API token
        """
        return cls(token = token)

    @classmethod
    def from_irondb(cls, endpoint, account=1):
        """
        Connect to an IRONdb node, instead of a CirconusAPI endpoint

        Args:
           - endpoint (str): IRONdb node URL, in the form "<protocol>://<hostname/ip>:<port>",
             e.g. "http://localhost:8112"
           - account (int): account id to use for CAQL requests.

        Notes:
           The current implementation will issue all requests against a single node.
        """
        return cls(endpoint = endpoint, account = account)

    def _caql_request(self, params):
        if self._mode == "API":
            return self._api.api_call("GET", "/caql", params=params)
        elif self._mode == "IRONdb":
            params = dict(params) # copy
            params['account_id'] = self._account
            resp = requests.post(
                self._endpoint + "/extension/lua/caql_v1",
                json=params
            )
            if resp.status_code == 200:
                return resp.json()
            else:
                raise Exception(resp.text)

    def caql(self, query, start, period, count, convert_hists = True, explain=False):
        """
        Fetch data using CAQL.

        Args:
           - query (str): the CAQL query string
           - start (int/datetime): starttime of the query.
             Either UNIX timestamp in seconds or datetime object
           - period (int): period of data to fetch
           - count (int): number of datapoints to fetch
           - convert_hists (boolean, optional): Convert returned histograms to
             Circllhist objects. Requires Circllhist to be available.

        Returns:
           res (dict): result in DF4 format. Example::

               {
                 "head" : { count = 60, start = ..., period = 60 }
                 "meta" : [ {kind:"numeric", label:"some_metric"}, {...}, ... ] -- per metric metadata
                 "data" : [ [1,1,1,1,...], [2,2,2,2,...] ] -- per metric data
               }
        """
        if isinstance(start, datetime):
            start = start.timestamp()
        if not start % period == 0:
            new_start = math.floor(start / period)
            warnings.warn(
                "start parameter {} is not divisible by period {}. Using {} instead.".format(
                    start, period, new_start))
            start = new_start

        params = {
            "explain" : explain,
            "query": query,
            "period": int(period),
            "start": int(start),
            "end": int(start + count * period),
            "format" : "DF4"
        }
        res = self._caql_request(params)

        # In the case of 0 output metrics, res['meta']/res['data'] might be None
        if not res['meta']: res['meta'] = []
        if not res['data']: res['data'] = []
        assert(len(res['meta']) == len(res['data']))

        if not convert_hists:
            return res

        #
        # Convert histogram JSON values to Circllhist objects.
        #
        if not Circllhist:
            raise ImportError("Circllhist not available")

        for i in range(len(res['meta'])):
            if res['meta'][i]['kind'] == "histogram":
                res['data'][i] = [ Circllhist.from_dict(h) for h in res['data'][i] ]

        return res

    def caqldf(self, *args, **kwargs):
        """
        Fetch CAQL as pandas DataFrame with ...

        - Columns : output metrics
        - Column names : metric labels
        - Row index : timestamps
        """
        if not pd:
            raise ImportError("pandas not available")
        res = self.caql(*args, **kwargs)
        head = res['head']
        meta = res['meta'] or []
        data = res['data'] or []
        return pd.DataFrame(
            data,
            columns = [
                datetime.fromtimestamp(head['start'] + i * head['period'])
                for i in range(head['count'])
            ],
            index = [ m['label'] for m in meta ],
        ).transpose()


circonusapi/circonusapi.py
"""
=================
Class CirconusAPI
=================

Example
-------
::

   from circonusapi import circonusapi

    # Initialize the API
    api = circonusapi.CirconusAPI(<api_token>)

    # List all check bundles
    checks = api.list_check_bundle()

    # Alternative (lower level) method
    api.api_call("GET", "/check_bundle")

    # Add a rule set
    data = {
        "check": "/check/12345",
        "contact_groups": {
            "1": ["On Call"],
            "2": [],
            "3": [],
            "4": [],
            "5": []
        },
        "derive": null,
        "link": null,
        "metric_name": "foo",
        "metric_type": "numeric",
        "notes": null,
        "parent": null,
        "rules": [
            {
                "criteria": "on absence",
                "severity": "1",
                "value": 1,
                "wait": "0"
            }
        ]
    }
    api.add_rule_set(data)

Methods
-------

All methods begin with a verb, then and underscore, and the circonus API
endpoint that you wish to operate on. For example, to view a check_bundle,
you would use api.get_check_bundle(1234), and to list all rule sets, use
api.list_rule_set().

The verbs/actions, and the associated HTTP methods are:

 * add - POST
 * edit - PUT
 * delete - DELETE
 * list - GET (without an ID specified)
 * get - GET (with an ID specified)

And the valid endpoints are (currently):

 * check_bundle
 * rule_set
 * graph
 * template
 * contact_group
 * broker
 * user
 * account


Method parameters
-----------------

Methods take two parameters where appropriate:

resource_id
   the ID of whatever you are fetching/modifying/deleting.
   This isn't required for list_X and add_X methods.

data
   Either a json string or python dict containing the data as required
   by the circonus API.


Direct Calls
------------

You can also make API calls directly without using methods. For example:

List check bundles::

    >>> api.api_call("GET", "/check_bundle")

Get a specific check bundle::

    >>> api.api_call("GET", "/check_bundle/1")

The direct method of making an API call is especially useful when
de-referencing a value you get back from the api. For example, the
_last_modified_by parameter is of the form '/user/XYZ', and you can find out
who this is by running api.api_call("GET", my_result['_last_modified_by']) and
not have to parse out the user number to use the regular api.get_user(123)
method.
"""

import json
import logging
import time

from contextlib import closing

try:
    from urllib.request import Request, urlopen
    from urllib.parse import quote, urlencode
    from urllib.error import HTTPError, URLError
    from http.client import IncompleteRead
except ImportError:
    # Python 2
    from urllib2 import Request, urlopen, HTTPError, URLError
    from urllib import quote, urlencode
    from httplib import IncompleteRead

log = logging.getLogger(__name__)


class CirconusAPI(object):
    """CirconusAPI Class"""

    def __init__(self, token, baseurl='https://api.circonus.com', appname='python-circonusapi',
                 debug=False):
        """Create a CirconusAPI object.

        Args:
           - token (str) : API token

        Kwargs:
           - baseurl (str) : URL of Circonus API endpoint to connect to.
           - appname (str) : Appname to use for authentification against the API
           - debug (boolean) : Turn on/off debugging

        """
        self.debug = False # Set api.debug = True to enable debug messages
        self.baseurl = baseurl
        self.appname = appname
        self.token = token
        self.endpoints = [
            'check_bundle',
            'rule_set',
            'rule_set_group',
            'graph',
            'template',
            'contact_group',
            'broker',
            'user',
            'account',
            'data',
            'maintenance',
            'alert',
            'annotation',
            'tag',
            'template'
        ]

        self.methods = {
            'add': {
                'method': 'POST',
                'id': False
            },
            'edit': {
                'method': 'PUT',
                'id': True
            },
            'delete': {
                'method': 'DELETE',
                'id': True
            },
            'list': {
                'method': 'GET',
                'id': False
            },
            'get': {
                'method': 'GET',
                'id': True
            }
        }

    def __getattr__(self, name):
        method, endpoint = name.split('_', 1)
        if method in self.methods and endpoint in self.endpoints:
            if self.methods[method]['id']:
                def f(resource_id=None, data=None, params=None):
                    return self.api_call(
                        self.methods[method]['method'],
                        "%s/%s" % (endpoint, resource_id),
                        data=data,
                        params=params
                    )
                return f
            else:
                def g(data=None, params={}):
                    return self.api_call(
                        self.methods[method]['method'],
                        endpoint,
                        data=data,
                        params=params
                    )
                return g
        else:
            raise AttributeError("%s instance has no attribute '%s'" % (
                self.__class__.__name__, name))

    def api_call(self, method, endpoint, data=None, params=None):
        """
        Performs a circonus api call.

        Args:
          - method (str) : HTTP Method, e.g. "GET" / "POST" / "DELETE"
          - endpoint (str) : Endpoint to request, e.g. "/checks"
          - data (str/dict) : Payload to send
          - params (dict) : Query string parameters

        """

        # Encode data as json if it isn't already. You can pass a json encoded
        # string or python dict here.
        if isinstance(data, dict):
            data = json.dumps(data)

        # converting str to bytes in PY3 or str to str PY2
        if data:
            data = data.encode('utf-8')

        # Allow specifying an endpoint both with and without a leading /
        endpoint = endpoint.lstrip('/')
        endpoint = quote(endpoint)
        if params:
            endpoint = '%s?%s' % (endpoint, urlencode(
                [(i, params[i]) for i in params]))
        url = "%s/v2/%s" % (self.baseurl, endpoint)
        req = Request(url=url, data=data,
            headers={
                "X-Circonus-Auth-Token": self.token,
                "X-Circonus-App-Name": self.appname,
                "Content-Type": "application/json",
                "Accept": "application/json"})
        req.get_method = lambda: method
        for i in range(5):
            # Retry 5 times until we succeed
            try:
                with closing(urlopen(req)) as fh:
                    response_data = fh.read().decode('utf-8')
                    code = fh.code
                    # We succeeded, exit the for loop
                    break
            except HTTPError as e:
                if e.code == 401:
                    raise TokenNotValidated
                if e.code == 403:
                    raise AccessDenied
                if e.code == 429:
                    # We got a rate limit error, retry
                    if self.debug:
                        log.debug("Rate limited. Retrying: %d", i)
                    time.sleep(1)
                    continue
                # Deal with other API errors
                try:
                    response_data = e.read().decode('utf-8')
                    e.close()
                    data_dict = json.loads(response_data)
                except ValueError:
                    data_dict = {}
                raise CirconusAPIError(e.code, data_dict, debug=self.debug)
            except (URLError, IncompleteRead):
                log.exception('Endpoint failed. Retrying. %s', url)
                time.sleep(1)
                continue
        else:
            # We have been rate limited, retried several times and still got
            # rate limited, so give up and raise an exception.
            raise RateLimitRetryExceeded()

        if self.debug:
            log.debug("data: %s", str(response_data))

        if code == 204:
            # Deal with empty response
            response = {}
        else:
            response = json.loads(response_data)
        # Deal with the unlikely case that we get an error with a 200 return
        # code
        if isinstance(response, dict) and not response.get('success', True):
            raise CirconusAPIError(200, response)
        return response


class CirconusAPIException(Exception):
    pass


class TokenNotValidated(CirconusAPIException):
    pass


class AccessDenied(CirconusAPIException):
    pass

class RateLimitRetryExceeded(CirconusAPIException):
    pass


class CirconusAPIError(CirconusAPIException):
    """Exception class for any errors thrown by the circonus API.

    Attributes:
        code -- the http code returned
        data -- the json object returned by the API
        success -- whether the request succeeded or failed (this is usually
                    false)
        error -- the error message returned by the API
    """
    def __init__(self, code, data, debug=False):
        self.code = code
        self.data = data
        self.debug = debug
        if hasattr(data, 'get'):
            self.success = data.get('success', False)
            self.message = data.get('message', '')
            self.explanation = data.get('explanation')
        else:
            self.message = str(data)

    def __str__(self):
        debuginfo = ""
        if self.debug:
            debuginfo = "\n%s" % json.dumps(self.data)
        return "HTTP %s - %s - %s%s" % (self.code, self.explanation,
                self.message, debuginfo)
